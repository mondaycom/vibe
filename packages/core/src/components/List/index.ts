export { default as List, type ListProps } from "./List";

export * from "./List.types";
