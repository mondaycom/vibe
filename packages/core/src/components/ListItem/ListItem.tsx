import cx from "classnames";
import React, { AriaAttributes, AriaRole, forwardRef, useCallback, useContext, useEffect, useRef } from "react";
import { camelCase } from "lodash-es";
import { getStyle } from "../../helpers/typesciptCssModulesHelper";
import Text from "../Text/Text";
import { SIZES, SELECTION_KEYS } from "../../constants";
import { NOOP } from "../../utils/function-utils";
import { withStaticProps, VibeComponentProps, VibeComponent, ElementContent } from "../../types";
import { useKeyEvent } from "../../hooks";
import useMergeRef from "../../hooks/useMergeRef";
import { ListContext } from "../List/utils/ListContext";
import { ListItemComponentType as ListItemComponentTypeEnum } from "./ListItemConstants";
import { ListItemElement, ListItemSize } from "./ListItem.types";
import styles from "./ListItem.module.scss";

export interface ListItemProps extends VibeComponentProps {
  /**
   * The HTML element used for the list item.
   */
  component?: ListItemElement;
  /**
   * The textual content inside the list item.
   */
  children?: ElementContent;
  /**
   * Callback fired when the item is clicked.
   */
  onClick?: (event: React.MouseEvent | React.KeyboardEvent, id: string) => void;
  /**
   * Callback fired when the item is hovered.
   */
  onHover?: (event: React.MouseEvent | React.FocusEvent, id: string) => void;
  /**
   * If true, disables the item and prevents interactions.
   */
  disabled?: boolean;
  /**
   * If true, marks the item as selected.
   */
  selected?: boolean;
  /**
   * The size of the list item.
   */
  size?: ListItemSize;
  /**
   * The tab index of the list item for keyboard navigation.
   */
  tabIndex?: number;
  /**
   * Indicates the current state of the item in a set of items.
   */
  "aria-current"?: AriaAttributes["aria-current"];
  /**
   * The ARIA role of the list item.
   */
  role?: AriaRole;
}

const ListItem: VibeComponent<ListItemProps> & { sizes?: typeof SIZES; components?: typeof ListItemComponentTypeEnum } =
  forwardRef(
    (
      {
        className,
        id,
        component = "div",
        onClick = NOOP,
        onHover = NOOP,
        selected,
        disabled = false,
        size = SIZES.SMALL,
        tabIndex = 0,
        children,
        "aria-current": ariaCurrent,
        "data-testid": dataTestId,
        role = "option"
      }: ListItemProps,
      ref
    ) => {
      const { updateFocusedItem } = useContext(ListContext);
      const componentRef = useRef(null);
      const mergedRef = useMergeRef(ref, componentRef);

      useEffect(() => {
        if (selected) {
          updateFocusedItem?.(id);
        }
      }, [selected, id, updateFocusedItem]);

      const componentOnClick = useCallback(
        (event: React.MouseEvent | React.KeyboardEvent) => {
          if (disabled) return;
          onClick(event, id);
        },
        [disabled, onClick, id]
      );

      useKeyEvent({
        keys: SELECTION_KEYS,
        ref: componentRef,
        callback: componentOnClick
      });

      const componentOnHover = useCallback(
        (event: React.MouseEvent | React.FocusEvent) => {
          if (disabled) return;
          onHover(event, id);
        },
        [disabled, onHover, id]
      );

      return (
        <Text
          element={component}
          data-testid={dataTestId || id}
          ref={mergedRef}
          className={cx(styles.listItem, className, getStyle(styles, camelCase(size)), {
            [styles.selected]: selected && !disabled,
            [styles.disabled]: disabled
          })}
          id={id}
          type="text2"
          aria-disabled={disabled}
          aria-selected={selected}
          onClick={componentOnClick}
          onMouseEnter={componentOnHover}
          onFocus={componentOnHover}
          role={role}
          tabIndex={tabIndex}
          aria-current={ariaCurrent}
        >
          {children}
        </Text>
      );
    }
  );

Object.assign(ListItem, {
  // Used by VirtualizedListItems
  displayName: "ListItem"
});

export default withStaticProps(ListItem, {
  sizes: SIZES,
  components: ListItemComponentTypeEnum
});
