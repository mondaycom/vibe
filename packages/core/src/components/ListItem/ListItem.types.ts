import { type SIZES } from "../../constants";

export type ListItemElement = "div" | "li" | "a";

export type ListItemSize = (typeof SIZES)[keyof typeof SIZES];
