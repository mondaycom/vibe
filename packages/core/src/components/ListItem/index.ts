export { default as ListItem, type ListItemProps } from "./ListItem";

export * from "./ListItem.types";
