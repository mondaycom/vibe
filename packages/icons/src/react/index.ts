/* eslint-disable */
/* tslint:disable */
export {default as Academy} from './Academy';
export {default as Activity} from './Activity';
export {default as Add} from './Add';
export {default as AddNewDoc} from './AddNewDoc';
export {default as AddSmall} from './AddSmall';
export {default as AddToTeam} from './AddToTeam';
export {default as AddUpdate} from './AddUpdate';
export {default as Alert} from './Alert';
export {default as AlignCenter} from './AlignCenter';
export {default as AlignLeft} from './AlignLeft';
export {default as AlignRight} from './AlignRight';
export {default as Announcement} from './Announcement';
export {default as API} from './API';
export {default as Apps} from './Apps';
export {default as Archive} from './Archive';
export {default as Attach} from './Attach';
export {default as Baseline} from './Baseline';
export {default as Basic} from './Basic';
export {default as Battery} from './Battery';
export {default as BlockQuote} from './BlockQuote';
export {default as Board} from './Board';
export {default as BoardPrivate} from './BoardPrivate';
export {default as BoardShareable} from './BoardShareable';
export {default as BoardTemplate} from './BoardTemplate';
export {default as Bold} from './Bold';
export {default as Bolt} from './Bolt';
export {default as Bookmark} from './Bookmark';
export {default as Broadcast} from './Broadcast';
export {default as Broom} from './Broom';
export {default as Bug} from './Bug';
export {default as Bullet} from './Bullet';
export {default as Bullets} from './Bullets';
export {default as Bulllet} from './Bulllet';
export {default as Calendar} from './Calendar';
export {default as Chart} from './Chart';
export {default as Check} from './Check';
export {default as Checkbox} from './Checkbox';
export {default as CheckList} from './CheckList';
export {default as ClassicFolder} from './ClassicFolder';
export {default as Clear} from './Clear';
export {default as Clipboard} from './Clipboard';
export {default as Close} from './Close';
export {default as CloseMedium} from './CloseMedium';
export {default as CloseRound} from './CloseRound';
export {default as CloseSmall} from './CloseSmall';
export {default as Code} from './Code';
export {default as Collapse} from './Collapse';
export {default as CollapseHorizontally} from './CollapseHorizontally';
export {default as CollapseRound} from './CollapseRound';
export {default as Column} from './Column';
export {default as Comment} from './Comment';
export {default as Completed} from './Completed';
export {default as Connect} from './Connect';
export {default as ConnectedDoc} from './ConnectedDoc';
export {default as ContentDirectory} from './ContentDirectory';
export {default as ConvertToItem} from './ConvertToItem';
export {default as ConvertToSubitem} from './ConvertToSubitem';
export {default as Counter} from './Counter';
export {default as Country} from './Country';
export {default as CreditCard} from './CreditCard';
export {default as Custom} from './Custom';
export {default as Cut} from './Cut';
export {default as Dashboard} from './Dashboard';
export {default as DashboardPrivate} from './DashboardPrivate';
export {default as Deactivate} from './Deactivate';
export {default as Delete} from './Delete';
export {default as Dependency} from './Dependency';
export {default as Description} from './Description';
export {default as DisabledUser} from './DisabledUser';
export {default as Divider} from './Divider';
export {default as Doc} from './Doc';
export {default as DocPrivate} from './DocPrivate';
export {default as DocShareable} from './DocShareable';
export {default as DocTemplate} from './DocTemplate';
export {default as DoubleCheck} from './DoubleCheck';
export {default as Downgrade} from './Downgrade';
export {default as Download} from './Download';
export {default as Drag} from './Drag';
export {default as Dropdown} from './Dropdown';
export {default as DropdownChevronDown} from './DropdownChevronDown';
export {default as DropdownChevronLeft} from './DropdownChevronLeft';
export {default as DropdownChevronRight} from './DropdownChevronRight';
export {default as DropdownChevronUp} from './DropdownChevronUp';
export {default as DueDate} from './DueDate';
export {default as Duplicate} from './Duplicate';
export {default as Edit} from './Edit';
export {default as Education} from './Education';
export {default as Email} from './Email';
export {default as Embed} from './Embed';
export {default as Emoji} from './Emoji';
export {default as Enter} from './Enter';
export {default as Erase} from './Erase';
export {default as Event} from './Event';
export {default as Expand} from './Expand';
export {default as ExternalPage} from './ExternalPage';
export {default as Favorite} from './Favorite';
export {default as Feedback} from './Feedback';
export {default as File} from './File';
export {default as Files} from './Files';
export {default as Filter} from './Filter';
export {default as Folder} from './Folder';
export {default as Form} from './Form';
export {default as Formula} from './Formula';
export {default as Forum} from './Forum';
export {default as Forward} from './Forward';
export {default as Fullscreen} from './Fullscreen';
export {default as FullscreenClose} from './FullscreenClose';
export {default as Gallery} from './Gallery';
export {default as Gantt} from './Gantt';
export {default as Gif} from './Gif';
export {default as Globe} from './Globe';
export {default as Graph} from './Graph';
export {default as Group} from './Group';
export {default as Guest} from './Guest';
export {default as Health} from './Health';
export {default as Heart} from './Heart';
export {default as Help} from './Help';
export {default as Hide} from './Hide';
export {default as Highlight} from './Highlight';
export {default as HighlightColorBucket} from './HighlightColorBucket';
export {default as Home} from './Home';
export {default as Idea} from './Idea';
export {default as Image} from './Image';
export {default as Inbox} from './Inbox';
export {default as Info} from './Info';
export {default as Integrations} from './Integrations';
export {default as Invite} from './Invite';
export {default as IPRestrictions} from './IPRestrictions';
export {default as Italic} from './Italic';
export {default as Item} from './Item';
export {default as ItemDefaultValues} from './ItemDefaultValues';
export {default as ItemHeightDouble} from './ItemHeightDouble';
export {default as ItemHeightSingle} from './ItemHeightSingle';
export {default as ItemHeightTriple} from './ItemHeightTriple';
export {default as Key} from './Key';
export {default as Keyboard} from './Keyboard';
export {default as Labs} from './Labs';
export {default as Launch} from './Launch';
export {default as Layout} from './Layout';
export {default as LearnMore} from './LearnMore';
export {default as Lines} from './Lines';
export {default as Link} from './Link';
export {default as Location} from './Location';
export {default as Locked} from './Locked';
export {default as LogIn} from './LogIn';
export {default as LogOut} from './LogOut';
export {default as LongText} from './LongText';
export {default as Mention} from './Mention';
export {default as Menu} from './Menu';
export {default as Microphone} from './Microphone';
export {default as Minimize} from './Minimize';
export {default as MinusSmall} from './MinusSmall';
export {default as Mirror} from './Mirror';
export {default as Mobile} from './Mobile';
export {default as MondayDoc} from './MondayDoc';
export {default as MondayLogoOutline} from './MondayLogoOutline';
export {default as Moon} from './Moon';
export {default as MoreActions} from './MoreActions';
export {default as MoreBelow} from './MoreBelow';
export {default as MoreBelowFilled} from './MoreBelowFilled';
export {default as Move} from './Move';
export {default as MoveArrowDown} from './MoveArrowDown';
export {default as MoveArrowLeft} from './MoveArrowLeft';
export {default as MoveArrowLeftDouble} from './MoveArrowLeftDouble';
export {default as MoveArrowLeftNarrow} from './MoveArrowLeftNarrow';
export {default as MoveArrowRight} from './MoveArrowRight';
export {default as MoveArrowRightNarrow} from './MoveArrowRightNarrow';
export {default as MoveArrowUp} from './MoveArrowUp';
export {default as Mute} from './Mute';
export {default as MyWeek} from './MyWeek';
export {default as NavigationChevronDown} from './NavigationChevronDown';
export {default as NavigationChevronLeft} from './NavigationChevronLeft';
export {default as NavigationChevronRight} from './NavigationChevronRight';
export {default as NavigationChevronUp} from './NavigationChevronUp';
export {default as NavigationDoubleChevronLeft} from './NavigationDoubleChevronLeft';
export {default as NewTab} from './NewTab';
export {default as Night} from './Night';
export {default as NoColor} from './NoColor';
export {default as Note} from './Note';
export {default as Notifications} from './Notifications';
export {default as NotificationsMuted} from './NotificationsMuted';
export {default as Numbers} from './Numbers';
export {default as Offline} from './Offline';
export {default as Open} from './Open';
export {default as Page} from './Page';
export {default as Paste} from './Paste';
export {default as Pause} from './Pause';
export {default as PDF} from './PDF';
export {default as Person} from './Person';
export {default as PersonRound} from './PersonRound';
export {default as Pin} from './Pin';
export {default as PinFull} from './PinFull';
export {default as Placeholder} from './Placeholder';
export {default as Play} from './Play';
export {default as Print} from './Print';
export {default as Prompt} from './Prompt';
export {default as PushNotification} from './PushNotification';
export {default as Quote} from './Quote';
export {default as Radio} from './Radio';
export {default as Recurring} from './Recurring';
export {default as RecycleBin} from './RecycleBin';
export {default as Redo} from './Redo';
export {default as Remove} from './Remove';
export {default as Reply} from './Reply';
export {default as ReplyAll} from './ReplyAll';
export {default as Retry} from './Retry';
export {default as Robot} from './Robot';
export {default as Rotate} from './Rotate';
export {default as ScheduledEmail} from './ScheduledEmail';
export {default as Search} from './Search';
export {default as Security} from './Security';
export {default as Send} from './Send';
export {default as Settings} from './Settings';
export {default as SettingsKnobs} from './SettingsKnobs';
export {default as Share} from './Share';
export {default as ShortText} from './ShortText';
export {default as Show} from './Show';
export {default as Shredder} from './Shredder';
export {default as Shuffle} from './Shuffle';
export {default as Sort} from './Sort';
export {default as SortAscending} from './SortAscending';
export {default as SortDescending} from './SortDescending';
export {default as Sound} from './Sound';
export {default as Status} from './Status';
export {default as StrikethroughS} from './StrikethroughS';
export {default as StrikethroughT} from './StrikethroughT';
export {default as Subitems} from './Subitems';
export {default as Sun} from './Sun';
export {default as Switch} from './Switch';
export {default as Switcher} from './Switcher';
export {default as Table} from './Table';
export {default as Tags} from './Tags';
export {default as Team} from './Team';
export {default as Text} from './Text';
export {default as TextBig} from './TextBig';
export {default as Textcolor} from './Textcolor';
export {default as TextColorIndicator} from './TextColorIndicator';
export {default as TextCopy} from './TextCopy';
export {default as TextHuge} from './TextHuge';
export {default as TextMedium} from './TextMedium';
export {default as TextSmall} from './TextSmall';
export {default as ThumbsDown} from './ThumbsDown';
export {default as ThumbsUp} from './ThumbsUp';
export {default as Time} from './Time';
export {default as Timeline} from './Timeline';
export {default as Translation} from './Translation';
export {default as TurnInto} from './TurnInto';
export {default as Underline} from './Underline';
export {default as Undo} from './Undo';
export {default as Unlocked} from './Unlocked';
export {default as Update} from './Update';
export {default as Upgrade} from './Upgrade';
export {default as Upload} from './Upload';
export {default as UserDomain} from './UserDomain';
export {default as UserStatus} from './UserStatus';
export {default as Versioning} from './Versioning';
export {default as Video} from './Video';
export {default as Wand} from './Wand';
export {default as Warning} from './Warning';
export {default as WarningFull} from './WarningFull';
export {default as WhatsNew} from './WhatsNew';
export {default as Widgets} from './Widgets';
export {default as Work} from './Work';
export {default as Workflow} from './Workflow';
export {default as Workspace} from './Workspace';
export {default as WrapText} from './WrapText';
/* tslint:enable */
/* eslint-enable */
