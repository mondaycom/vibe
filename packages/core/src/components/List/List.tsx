import cx from "classnames";
import React, {
  AriaAttributes,
  AriaRole,
  CSSProperties,
  forwardRef,
  ReactElement,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState
} from "react";
import useMergeRef from "../../hooks/useMergeRef";
import useKeyEvent from "../../hooks/useKeyEvent";
import { VirtualizedListItems } from "./VirtualizedListItems/VirtualizedListItems";
import { keyCodes, UP_DOWN_ARROWS } from "../../constants/keyCodes";
import { VibeComponent, withStaticProps, VibeComponentProps } from "../../types";
import { ListItemProps } from "../ListItem/ListItem";
import { ListTitleProps } from "../ListTitle/ListTitle";
import { ListWrapperComponentType as ListWrapperComponentTypeEnum } from "./ListConstants";
import { ListElement } from "./List.types";
import { ComponentDefaultTestId, getTestId } from "../../tests/test-ids-utils";
import { ListContext } from "./utils/ListContext";
import {
  getListItemComponentType,
  getListItemIdByIndex,
  getListItemIndexById,
  getNextListItemIndex,
  getPrevListItemIndex,
  isListItem,
  useListId
} from "./utils/ListUtils";
import styles from "./List.module.scss";
import { ComponentVibeId } from "../../tests/constants";

export interface ListProps extends VibeComponentProps {
  /**
   * the wrapping component to wrap the List Items [div, nav, ul, ol]
   */
  component?: ListElement;
  /**
   * ARIA label string to describe to list
   */
  ariaLabel?: string;
  /**
   * ARIA described by string to reference an id to describe by
   */
  ariaDescribedBy?: string;
  "aria-controls"?: AriaAttributes["aria-controls"];
  children?: ReactElement<ListItemProps | ListTitleProps> | ReactElement<ListItemProps | ListTitleProps>[];
  /**
   * Using virtualized list for rendering only the items which visible to the user in any given user (performance optimization)
   */
  renderOnlyVisibleItems?: boolean;
  style?: CSSProperties;
  /**
   * ARIA role for the list.
   */
  role?: AriaRole;
}

const List: VibeComponent<ListProps> & {
  components?: typeof ListWrapperComponentTypeEnum;
} = forwardRef(
  (
    {
      className,
      id,
      component = List.components.UL,
      children,
      ariaLabel,
      ariaDescribedBy,
      "aria-controls": ariaControls,
      renderOnlyVisibleItems = false,
      style,
      role = "listbox",
      "data-testid": dataTestId
    }: ListProps,
    ref
  ) => {
    const componentRef = useRef(null);
    const mergedRef = useMergeRef(ref, componentRef);
    const childrenRefs: React.MutableRefObject<HTMLElement[]> = useRef([]);

    const overrideId = useListId(id);
    const [focusIndex, setFocusIndex] = useState(0);
    const Component = component;

    const updateFocusedItem = useCallback((id: string) => {
      setFocusIndex(getListItemIndexById(childrenRefs, id));
      componentRef?.current?.setAttribute("aria-activedescendant", id);
    }, []);

    const onUpDownArrows = useCallback(
      (event: KeyboardEvent) => {
        if (renderOnlyVisibleItems) {
          return;
        }
        event.preventDefault();

        const isUpKey = event.key === keyCodes.UP_ARROW;
        const isDownKey = event.key === keyCodes.DOWN_ARROW;
        let overrideFocusIndex = undefined;

        if (isDownKey && focusIndex + 1 < childrenRefs.current.length) {
          overrideFocusIndex = getNextListItemIndex(focusIndex, childrenRefs);
        } else if (isUpKey && focusIndex > 0) {
          overrideFocusIndex = getPrevListItemIndex(focusIndex, childrenRefs);
        }
        if (overrideFocusIndex !== undefined) {
          updateFocusedItem(getListItemIdByIndex(childrenRefs, overrideFocusIndex));
          childrenRefs.current[overrideFocusIndex].focus();
        }
      },
      [focusIndex, renderOnlyVisibleItems, updateFocusedItem]
    );

    useKeyEvent({
      keys: UP_DOWN_ARROWS,
      callback: onUpDownArrows,
      ref: componentRef
    });

    useEffect(() => {
      const selectedItemIndex = childrenRefs.current.findIndex(
        child => child instanceof HTMLElement && isListItem(child) && child?.getAttribute("aria-selected") === "true"
      );
      if (selectedItemIndex !== -1) {
        updateFocusedItem(getListItemIdByIndex(childrenRefs, selectedItemIndex));
      } else {
        const firstFocusableIndex = childrenRefs.current.findIndex(child => isListItem(child));
        if (firstFocusableIndex !== -1) {
          updateFocusedItem(getListItemIdByIndex(childrenRefs, firstFocusableIndex));
        }
      }
    }, [updateFocusedItem]);

    const overrideChildren = useMemo(() => {
      let override: ReactElement | ReactElement[] = Array.isArray(children) ? children : [children];
      if (renderOnlyVisibleItems) {
        override = <VirtualizedListItems>{override}</VirtualizedListItems>;
      } else {
        childrenRefs.current = childrenRefs.current.slice(0, override.length);
        override = React.Children.map(override, (child, index) => {
          if (!React.isValidElement(child)) {
            return child;
          }
          const id = (child.props as ListItemProps).id || `${overrideId}-item-${index}`;
          const currentRef = childrenRefs.current[index];
          const isFocusableItem = currentRef === undefined || currentRef === null || isListItem(currentRef);
          return React.cloneElement(child, {
            // @ts-ignore not sure how to deal with ref here
            ref: ref => (childrenRefs.current[index] = ref),
            tabIndex: focusIndex === index && isFocusableItem ? 0 : -1,
            id,
            component: getListItemComponentType(component),
            role: (child.props as ListItemProps).role
          });
        });
      }

      return override;
    }, [children, component, focusIndex, overrideId, renderOnlyVisibleItems]);

    return (
      <ListContext.Provider value={{ updateFocusedItem }}>
        <Component
          data-testid={dataTestId || getTestId(ComponentDefaultTestId.LIST, id)}
          data-vibe={ComponentVibeId.LIST}
          ref={mergedRef}
          style={style}
          className={cx(styles.list, className)}
          id={overrideId}
          aria-label={ariaLabel}
          aria-describedby={ariaDescribedBy}
          aria-controls={ariaControls}
          tabIndex={-1}
          role={role}
        >
          {overrideChildren}
        </Component>
      </ListContext.Provider>
    );
  }
);

export default withStaticProps(List, {
  components: ListWrapperComponentTypeEnum
});
