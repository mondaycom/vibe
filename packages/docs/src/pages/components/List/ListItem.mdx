import { Canvas, Meta } from "@storybook/blocks";
import * as ListItemStories from "./ListItem.stories";

<Meta of={ListItemStories} />

# ListItem

An item of a List component

<Canvas of={ListItemStories.Overview} />

### Import

```js
import { ListItem, ListItemIcon, ListItemAvatar } from "@vibe/core";
```

## Props

<PropsTable />

## Usage

<UsageGuidelines
  guidelines={[
    "List item can include an ListItemIcon or text.",
    "List item icon should always be displayed before the item's text (left or right depending on the item's text language)"
  ]}
/>

## Variants

### States

<Canvas of={ListItemStories.States} />

### Sizes

<Canvas of={ListItemStories.Sizes} />

### List item with an icon

<Canvas of={ListItemStories.WithIcon} />

### List item with an avatar

<Canvas of={ListItemStories.WithAvatar} />
