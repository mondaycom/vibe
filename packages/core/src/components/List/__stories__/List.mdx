import List from "../List";
import { Meta } from "@storybook/blocks";
import { Board } from "@vibe/icons";
import ListItem from "../../ListItem/ListItem";
import ListItemIcon from "../../ListItemIcon/ListItemIcon";
import {
  MENU,
  MULTI_STEP_INDICATOR,
  TABS
} from "../../../storybook/components/related-components/component-description-map";
import { Filter, Add } from "@vibe/icons";
import classes from "./List.stories.module.scss";
import Divider from "../../Divider/Divider";
import DialogContentContainer from "../../DialogContentContainer/DialogContentContainer";
import { Link, Tip } from "vibe-storybook-components";
import { TipCheckYourself } from "./List.stories.helpers";
import * as ListStories from "./List.stories";

<Meta of={ListStories} />

# List

- [Overview](#overview)
- [Props](#props)
- [Usage](#usage)
- [Variants](#variants)
- [Do’s and don’ts](#dos-and-donts)
- [Related components](#related-components)
- [Feedback](#feedback)

## Overview

Lists is a group of actionable items containing primary and supplemental actions, which are represented by icons and text.

<Canvas of={ListStories.Overview} />

### Import

```js
import { List } from "@vibe/core";
```

## Props

<PropsTable />

## Usage

<UsageGuidelines
  guidelines={[
    "Use list component for display a list of navigable items in a column.",
    "List purpose is usually allowing navigation in context to the navigable  content.",
    "Your list items can be either in regular, selected or disabled state. You can control a list item's component state by its attributes."
  ]}
/>

<TipCheckYourself />

## Variants

### List with categories

<Canvas of={ListStories.ListWithCategories} />

### List with icons

<Canvas of={ListStories.ListWithIcons} />

### List with avatars

<Canvas of={ListStories.ListWithAvatars} />

### List with virtualization optimization

When you display a large number of items, you may want to render only the options shown at a given moment to allow better performance and a
better user experience.

<Canvas of={ListStories.ListWithVirtualizationOptimization} />

<Tip>
  While using virtualization optimization on your list component, please be aware your list needs to contain a constant
  height and width for function properly
</Tip>

## Do’s and Don’ts

<ComponentRules
  rules={[
    {
      positive: {
        component: (
          <div className={classes["leftpane-workspaces-section"]}>
            <List>
              <ListItem>
                <ListItemIcon icon={Add} />
                Add
              </ListItem>
              <ListItem>
                <ListItemIcon icon={Filter} />
                Filter
              </ListItem>
            </List>
            <Divider />
            <List>
              <ListItem>
                <ListItemIcon icon={Board} />
                Board name
              </ListItem>
              <ListItem>
                <ListItemIcon icon={Board} />
                Board name
              </ListItem>
            </List>
          </div>
        ),
        description: "Use a list component to create a menu displayed on one of the various sections of your web page."
      },
      negative: {
        component: (
          <DialogContentContainer>
            <List>
              <ListItem>
                <ListItemIcon icon={Add} />
                Add
              </ListItem>
              <ListItem>
                <ListItemIcon icon={Filter} />
                Filter
              </ListItem>
            </List>
            <Divider />
            <List>
              <ListItem>
                <ListItemIcon icon={Board} />
                Board name
              </ListItem>
              <ListItem>
                <ListItemIcon icon={Board} />
                Board name
              </ListItem>
            </List>
          </DialogContentContainer>
        ),
        description: (
          <>
            Do not use the list to implement a menu displayed inside a dialog. See{" "}
            <StorybookLink page="Components/Menu">Menu</StorybookLink>.
          </>
        )
      }
    }
  ]}
/>

## Related components

<RelatedComponents componentsNames={[MENU, TABS, MULTI_STEP_INDICATOR]} />
