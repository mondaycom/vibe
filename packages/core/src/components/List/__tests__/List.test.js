import React from "react";
import renderer from "react-test-renderer";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import List from "../List";
import ListItem from "../../ListItem/ListItem";

describe("List", () => {
  it("renders correctly without props", () => {
    const tree = renderer.create(<List />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly with empty props", () => {
    const tree = renderer.create(<List>Something</List>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly with list items", () => {
    const tree = renderer
      .create(
        <List>
          <ListItem>1</ListItem>
          <ListItem>2</ListItem>
        </List>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe("accessibility tests", () => {
    it("List should have role listbox", () => {
      const { getByRole } = render(
        <List>
          <ListItem>1</ListItem>
          <ListItem>2</ListItem>
        </List>
      );
      expect(getByRole("listbox")).toBeInTheDocument();
    });

    it("ListItem should have role option", () => {
      const { getAllByRole } = render(
        <List>
          <ListItem>1</ListItem>
          <ListItem>2</ListItem>
        </List>
      );
      expect(getAllByRole("option")).toHaveLength(2);
    });

    it("disabled ListItem should have aria-selected", () => {
      const { getByTestId } = render(
        <List>
          <ListItem data-testid="list-item-1" disabled>
            1
          </ListItem>
          <ListItem data-testid="list-item-2">1</ListItem>
        </List>
      );
      expect(getByTestId("list-item-1")).toHaveAttribute("aria-disabled", "true");
      expect(getByTestId("list-item-2")).toHaveAttribute("aria-disabled", "false");
    });

    it("selected ListItem should have aria-selected", () => {
      const { getByTestId } = render(
        <List>
          <ListItem data-testid="list-item-1" selected>
            1
          </ListItem>
          <ListItem data-testid="list-item-2">1</ListItem>
        </List>
      );
      expect(getByTestId("list-item-1")).toHaveAttribute("aria-selected", "true");
      expect(getByTestId("list-item-2")).not.toHaveAttribute("aria-selected");
    });

    it("List should have aria-activedescendant", () => {
      const { getByRole } = render(
        <List>
          <ListItem id="list-item-1">1</ListItem>
          <ListItem id="list-item-2" selected>
            2
          </ListItem>
        </List>
      );
      expect(getByRole("listbox")).toHaveAttribute("aria-activedescendant", "list-item-2");
    });

    it("List aria-activedescendant", () => {
      const { getByRole } = render(
        <List id="list">
          <ListItem>1</ListItem>
          <ListItem>2</ListItem>
          <ListItem selected>3</ListItem>
        </List>
      );
      const list = getByRole("listbox");
      expect(list).toHaveAttribute("aria-activedescendant", "list-item-2");
      userEvent.tab();
      userEvent.keyboard("{arrowup}");
      expect(list).toHaveAttribute("aria-activedescendant", "list-item-1");
    });
  });

  describe("custom roles", () => {
    it("List render with a custom role", () => {
      const { getByRole, getAllByRole } = render(
        <List role="list">
          <ListItem role="listitem">1</ListItem>
          <ListItem role="listitem">2</ListItem>
        </List>
      );
      expect(getByRole("list")).toBeInTheDocument();
      expect(getAllByRole("listitem")).toHaveLength(2);
    });

    it("selected ListItem should have aria-selected with custom roles", () => {
      const { getByTestId } = render(
        <List role="list">
          <ListItem data-testid="list-item-1" role="listitem" selected>
            1
          </ListItem>
          <ListItem data-testid="list-item-2" role="listitem">
            2
          </ListItem>
        </List>
      );
      expect(getByTestId("list-item-1")).toHaveAttribute("aria-selected", "true");
      expect(getByTestId("list-item-2")).not.toHaveAttribute("aria-selected");
    });

    it("List should have aria-activedescendant with custom roles", () => {
      const { getByRole } = render(
        <List role="list">
          <ListItem id="list-item-1" role="listitem">
            1
          </ListItem>
          <ListItem id="list-item-2" role="listitem" selected>
            2
          </ListItem>
        </List>
      );
      expect(getByRole("list")).toHaveAttribute("aria-activedescendant", "list-item-2");
    });

    it("List aria-activedescendant with custom roles", () => {
      const { getByRole } = render(
        <List id="list" role="list">
          <ListItem role="listitem">1</ListItem>
          <ListItem role="listitem">2</ListItem>
          <ListItem role="listitem" selected>
            3
          </ListItem>
        </List>
      );
      const list = getByRole("list");
      expect(list).toHaveAttribute("aria-activedescendant", "list-item-2");
      userEvent.tab();
      userEvent.keyboard("{arrowup}");
      expect(list).toHaveAttribute("aria-activedescendant", "list-item-1");
    });
  });
});
