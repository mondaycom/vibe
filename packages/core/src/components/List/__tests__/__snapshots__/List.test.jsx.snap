// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`List > renders correctly with empty props 1`] = `
<ul
  className="list"
  data-testid="list"
  data-vibe="List"
  id="list-1"
  role="listbox"
  tabIndex={-1}
>
  Something
</ul>
`;

exports[`List > renders correctly with list items 1`] = `
<ul
  className="list"
  data-testid="list"
  data-vibe="List"
  id="list-2"
  role="listbox"
  tabIndex={-1}
>
  <li
    aria-disabled={false}
    className="typography primary start singleLineEllipsis text text2Normal listItem small"
    data-testid="list-2-item-0"
    id="list-2-item-0"
    onClick={[Function]}
    onFocus={[Function]}
    onMouseEnter={[Function]}
    role="option"
    tabIndex={0}
  >
    1
  </li>
  <li
    aria-disabled={false}
    className="typography primary start singleLineEllipsis text text2Normal listItem small"
    data-testid="list-2-item-1"
    id="list-2-item-1"
    onClick={[Function]}
    onFocus={[Function]}
    onMouseEnter={[Function]}
    role="option"
    tabIndex={-1}
  >
    2
  </li>
</ul>
`;

exports[`List > renders correctly without props 1`] = `
<ul
  className="list"
  data-testid="list"
  data-vibe="List"
  id="list-0"
  role="listbox"
  tabIndex={-1}
/>
`;
